src/include/btmem.h

```diff
@@ -1131,8 +1132,33 @@ struct __wt_update {
      */
     volatile uint8_t prepare_state; /* prepare state */
 
+/*
+ * WT_UPDATE_DS indicates that an update is intended to be written to the data store during a round
+ * of reconciliation. If the flag is set on an update it indicates that subsequent rounds of
+ * reconciliation can write that update even if they can't see it. This is done to prevent an older
+ * update in the update chain overwriting the newer on-page update. Ideally this flag would only
+ * be set if we knew that the reconciliation would be successful, however we cannot know that.  As
+ * such the reconciliation can still fail after setting the flag, this is okay. However be extremely
+ * careful when adding logic to the code that uses this flag. Generally the reconciliation will only
+ * fail when a mixed-mode or out-of-order timestamp update is encountered by eviction and a
+ * checkpoint is running concurrently.
+ *
+ * It is only flagged on the on-page update if ALL older updates were successfully wrote to the
+ * history store in the history store insert path. It may also be flagged on the tombstone
+ * preceding the on-page update if one exists in the same code path.
+ *
+ * There are two other scenarios where we could have marked the on page update as required to be
+ * written to the datastore, they are being listed here for completeness:
+ *  - When we select a tombstone to write to the data store, in this case no entries are
+ *    written to the history store as such it would need to be set in the rec_row path.
+ *    However because the tombstone is globally visible no other round of reconciliation
+ *    would choose an update before it.
+ *  - When we an update on a key that was just inserted. As such no updates are behind it
```

Comment:
typo

Reply:
We still need to flag it right? Suppose checkpoint comes to it first and write that value but we don't flag it. The next eviction comes and can't see the value, then the key is deleted in the eviction disk image.

Reply:
I think it is OK as there is no history store value but the inconsistency worries me.

Reply:
No I don't think the delete case is okay but if we have this as a concern then its worth discussing on a call.

Reply:
Given the discussion on the call this should be okay but isn't ideal. There are two cases to consider:

**1. Checkpoint ahead of eviction**
If checkpoint is ahead of eviction and selects this update as the on page update the disk image will be wrote containing it. Later if eviction comes to the page and can't see this update it'll write a page image without the update and the update will persist in memory as eviction thinks it's uncommitted. If we were to crash here we'd be okay as we'd load the checkpoint which still contains the key in the page image.

**2. Eviction ahead of checkpoint**
If eviction is ahead of checkpoint and evicts the page, the page image on disk contains the key/value. A checkpoint begins, and an update is committed to the key that checkpoint can't see, dirtying the page. Checkpoint visits this update chain and should restore the update flagging it as WT_UPDATE_RESTORED_FROM_DS and then will chose to write it.